src/rebuild/srv.c

```diff
@@ -866,26 +870,69 @@ rebuild_prepare(struct ds_pool *pool, uint32_t rebuild_ver,
  * rebuild.
  */
 static int
-rebuild_scan_broadcast(struct ds_pool *pool,
-		       struct rebuild_global_pool_tracker *rgt,
-		       struct pool_target_id_list *tgts_failed,
+rebuild_scan_broadcast(struct ds_pool *pool, struct rebuild_global_pool_tracker *rgt,
+		       struct pool_target_id_list *tgts_failed, uint32_t max_ver,
 		       daos_rebuild_opc_t rebuild_op)
 {
 	struct rebuild_scan_in	*rsi;
 	struct rebuild_scan_out	*rso;
+	d_rank_list_t		*excluded = NULL;
 	crt_rpc_t		*rpc;
 	int			rc;
 
-	/* Send rebuild RPC to all targets of the pool to initialize rebuild.
-	 * XXX this should be idempotent as well as query and fini.
+	/* There might be some other ranks being queued for reintegration,
+	 * but not included in this reintegration, so let's exclude those
+	 * ranks from this reintegration.
 	 */
+	D_DEBUG(DB_REBUILD, "rebuild op %d\n", rebuild_op);
+	if (rebuild_op == RB_OP_REINT || rebuild_op == RB_OP_EXTEND ||
+	    rebuild_op == RB_OP_RECLAIM) {
+		d_rank_list_t	up_ranks = { 0 };
+		int		i;
+		int		nr = 0;
+
+		ABT_rwlock_rdlock(pool->sp_lock);
+		rc = map_ranks_init(pool->sp_map, PO_COMP_ST_UP, &up_ranks);
+		ABT_rwlock_unlock(pool->sp_lock);
+		if (rc != 0) {
+			D_ERROR(DF_UUID": failed to create rank list: %d\n",
+				DP_UUID(pool->sp_uuid), rc);
+			return rc;
+		}
+
+		D_DEBUG(DB_REBUILD, "up_ranks %d\n", up_ranks.rl_nr);
+		excluded = d_rank_list_alloc(up_ranks.rl_nr);
```

Comment:
is it possible the up_ranks.rl_nr == 0? need to handle that case?

Reply:
hmm, even if UP ranks are empty, this will also work. though this can be optimized a bit.

Reply:
right, thx